### DISCUSSION TITLE
Grad & Div

### DISCUSSION THREAD

**Discussion by mscsiro** [sintefmath/JutulDarcy.jl#70] [Q&A]
Hi- I'm newbie to JutulDarcy but have been using MRST for a while, especially for prototyping (like adding custom physics). As expected, I’m hitting some performance limits with MRST, so I’m hoping to switch to JutulDarcy when performance becomes a concern. I learned MRST using this tutorial: (https://folk.ntnu.no/andreas/hasselt17/introMRST.pdf), and I was wondering if there's anything similar for JutulDarcy. Specifically, I’m looking for guidance on how to initialise AD variables (like initVariablesADI in MRST) and use operators like Grad and Div to add more physics. Right now, I’m working on a code in MRST where I’ve added sorption/diffusion, but I’m struggling to get started with JutulDarcy. Any help would be appreciated!

**Comment by moyner** [comment]
Hi! Apologies for the late reply.

JutulDarcy is a bit complex when coming from MRST, but it should be possible to do the modifications here as well.

The key idea is that JutulDarcy works with types, and you add additional function overloads to get the results you want. Here are a few examples from Jutul.jl that are examples of prototyping:

https://github.com/sintefmath/Jutul.jl/blob/07c74861eb518c1d2b18a146e7c3c62651ec517d/src/applications/test_systems/variable_poisson/variable_poisson.jl#L108

https://sintefmath.github.io/Jutul.jl/dev/#An-example

For modifying the reservoir simulator, it depends a bit on the kind of physics you want to add. Do you have an example of the kind of equations you are working with? The short answer is that JutulDarcy automatically takes care of Div - you have to write the implementation of the value that goes in the accumulation term and the "face flux" for a given component.

One place to look at how this is done is the thermal equations, that get added as an additional set of energy conservation equations to an existing flow system: https://github.com/sintefmath/JutulDarcy.jl/tree/main/src/thermal

Best regards,
Olav

**Reply by mscsiro** [reply]
Thank you, Olav, for the reply and the links. While I go through the examples, I thought I’d elaborate more on the current problem I’m trying to translate from MRST to JutulDarcy. It involves a custom multicomponent dual-porosity model, where the matrix and fracture are linked through a sink/source term described by diffusion (a concentration gradient). To achieve this, I need to duplicate the equations (one for the matrix and one for the fracture), couple them through the sink/source term, and introduce a new set of primary variables (the concentrations), using AD.  

Fracture mass con.: 
\frac{\partial (\phi c_i)}{\partial t} + \nabla \cdot \mathbf{J}_i = q_i + **q_i^{mf}**, \quad i = 1, 2, \dots, N
Matrix mass con.:
\frac{\partial m_i}{\partial t} = -**q_i^{mf}**, \quad i = 1, 2, \dots, N

I reckon I can't (or better to say dont know how to) use Jutul directly, as I need some of JutulDarcy's functionalities, including the EoS (for density and viscosity calculations), grid processing (for corner-point grids), and possibly the well models from JutulDarcy.
 
 I would like to start from an easier example though, for instance adding a dispersion term to a single-porosity model and then build on that. By debugging and going step by step through the codes using the examples, I tried to comprehend the process and get to the equations; however as you know the debugging using VS code for Julia is not very smooth and fast, unlike MatLab or Python. So,  
 
 Cheers,
